### PR TITLE
improve cache

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why this change is needed

Many values stored in the Ten Gateway's caching layer expire as soon as a new batch is created.
This PR adds this feature.

### What changes were made as part of this PR

- change the CacheConfig to return some cache strategy types instead of durations
- implement a cache eviction check 

### PR checks pre-merging

Please indicate below by ticking the checkbox that you have read and performed the required
[PR checks](https://github.com/ten-protocol/ten-internal/blob/main/dev-ops-docs/dev-pr-checks.md)

- [x] PR checks reviewed and performed

### CODE REVIEW

File: tools/walletextension/cache/RistrettoCache.go [hunk 1]
```diff
@@ -43,6 +45,21 @@ func NewRistrettoCache(logger log.Logger) (Cache, error) {
 	return c, nil
 }
 
+func (c *ristrettoCache) EvictShortLiving() {
+	c.lastEviction = time.Now()
+}
+
+func (c *ristrettoCache) IsEvicted(key any) bool {
```
Comment (tudor-malene) [inline tools/walletextension/cache/RistrettoCache.go:52]: This here is the fiddly bit.
The rest is relatively straight forward

File: tools/walletextension/rpcapi/wallet_extension.go [hunk 2]
```diff
@@ -118,7 +118,7 @@ func NewServices(hostAddrHTTP string, hostAddrWS string, storage storage.Storage
 
 	services.backendNewHeadsSubscription = clientSubscription
 	services.NewHeadsService = subscriptioncommon.NewNewHeadsService(ch, true, logger, func(newHead *common2.BatchHeader) error {
-		// todo - in a followup PR, invalidate cache entries marked as "latest"
+		services.Cache.EvictShortLiving()
```
Comment (BedrockSquirrel) [inline tools/walletextension/rpcapi/wallet_extension.go:121]: Missed this on a previous review but could you alias that `common2` import while you're here?

File: tools/walletextension/cache/RistrettoCache.go [hunk 3]
```diff
@@ -43,6 +45,21 @@ func NewRistrettoCache(logger log.Logger) (Cache, error) {
...
+	remainingTTL, notExpired := c.cache.GetTTL(key)
+	if !notExpired {
+		return true
+	}
+	cachedTime := time.Now().Add(-remainingTTL)
```
[the remark this answers is not in the document]
Comment (tudor-malene) [reply tools/walletextension/cache/RistrettoCache.go:57]: you're right. I got confused.

File: tools/walletextension/rpcapi/utils.go [hunk 4]
```diff
@@ -44,12 +44,17 @@ type ExecCfg struct {
...
+)
+
 type CacheCfg struct {
-	// ResetWhenNewBlock bool todo
-	TTL time.Duration
-	// logic based on block
-	// todo - handle block in the future
-	TTLCallback func() time.Duration
+	CacheType        CacheStrategy
+	CacheTypeDynamic func() CacheStrategy
```
[the remark this answers is not in the document]
Comment (tudor-malene) [reply tools/walletextension/rpcapi/utils.go:57]: I tried to keep things straight-forward and compact where you know exactly

File: tools/walletextension/cache/RistrettoCache.go [hunk 5]
```diff
@@ -49,12 +49,12 @@ func (c *ristrettoCache) EvictShortLiving() {
 	c.lastEviction = time.Now()
 }
 
-func (c *ristrettoCache) IsEvicted(key any) bool {
+func (c *ristrettoCache) IsEvicted(key any, ttl time.Duration) bool {
```
Comment (BedrockSquirrel) [inline tools/walletextension/cache/RistrettoCache.go:52]: Minor: could maybe call this `originalTTL` for clarity.

**Review by BedrockSquirrel — APPROVED** [review]
Nice, LGTM

**Pull request merged**